### PR TITLE
[TIMOB-19383] iOS9: Upgrade Facebook module to facebook 4.7.0

### PR DESCRIPTION
**Pull request by cheekiatng**
JIRA: https://jira.appcelerator.org/browse/TIMOB-19383

### CODE REVIEW

File: apidoc/LikeButton.yml [hunk 1]
```diff
@@ -115,6 +114,16 @@ properties:
     description: |
         This property must be set before rendering the button.
     type: String
+    deprecated:
+        since: 5.0.0
```
Comment (hansemannn) [inline apidoc/LikeButton.yml:118]: Please add the "removed: 5.0.0" line here, as it says it is removed in the description.

File: README.md [hunk 2]
```diff
@@ -195,23 +235,12 @@ To share more information, example:
 
 ```javascript
     var fb = require('facebook');
-    if(fb.getCanPresentShareDialog()) { //checks to see if facebook app installed
```
Comment (hansemannn) [inline README.md:198]: It this method not needed anymore and the fallback is presented automatically?

File: ios/Classes/FacebookLikeButton.m [hunk 3]
```diff
@@ -33,8 +32,14 @@ -(void)frameSizeChanged:(CGRect)frame bounds:(CGRect)bounds
 
 -(void)setObjectId_:(id)idString
 {
+    DEPRECATED_REPLACED(@"facebook.likeButton.objectId", @"5.0.0", @"facebook.likeButton.objectId")
```
Comment (hansemannn) [inline ios/Classes/FacebookLikeButton.m]: Capitalize "facebook" to "Facebook".

File: ios/Classes/FacebookLikeButton.m [hunk 4]
```diff
@@ -50,63 +55,56 @@ -(void)setForegroundColor_:(id)colorVal
...
     ENSURE_SINGLE_ARG(styleStr, NSString);
-    FBLikeControl *btn = [self like];
+    FBSDKLikeControl *btn = [self like];
     NSString* style = [TiUtils stringValue:styleStr];
     if ([style isEqualToString:@"box_count"]) {
-        btn.likeControlStyle = FBLikeControlStyleBoxCount;
+        btn.likeControlStyle = FBSDKLikeControlStyleBoxCount;
     } else if ([style isEqualToString:@"button"]) {
-        btn.likeControlStyle = FBLikeControlStyleButton;
+        DEPRECATED_REMOVED(@"facebook.likeButton.likeViewStyle.button", @"5.0.0", @"5.0.0");
```
Comment (hansemannn) [inline ios/Classes/FacebookLikeButton.m]: Capitalize "facebook" to "Facebook".

File: ios/Classes/FacebookLikeButton.m [hunk 5]
```diff
@@ -50,63 +55,56 @@ -(void)setForegroundColor_:(id)colorVal
...
     }
 }
 
 -(void)setAuxiliaryViewPosition_:(id)positionStr
 {
     ENSURE_SINGLE_ARG(positionStr, NSString);
-    FBLikeControl *btn = [self like];
+    FBSDKLikeControl *btn = [self like];
     NSString* position = [TiUtils stringValue:positionStr];
     if ([position isEqualToString:@"bottom"]) {
```
Comment (hansemannn) [inline ios/Classes/FacebookLikeButton.m:74]: The attributes "inline", "top" and "bottom" not seem to be documented.

File: ios/Classes/FacebookLikeButton.m [hunk 6]
```diff
@@ -50,63 +55,56 @@ -(void)setForegroundColor_:(id)colorVal
...
     if ([align isEqualToString:@"center"]) {
-        btn.likeControlHorizontalAlignment = FBLikeControlHorizontalAlignmentCenter;
+        btn.likeControlHorizontalAlignment = FBSDKLikeControlHorizontalAlignmentCenter;
     } else if ([align isEqualToString:@"left"]) {
-        btn.likeControlHorizontalAlignment = FBLikeControlHorizontalAlignmentLeft;
+        btn.likeControlHorizontalAlignment = FBSDKLikeControlHorizontalAlignmentLeft;
     } else if ([align isEqualToString:@"right"]) {
-        btn.likeControlHorizontalAlignment = FBLikeControlHorizontalAlignmentRight;
+        btn.likeControlHorizontalAlignment = FBSDKLikeControlHorizontalAlignmentRight;
     }
```
Comment (hansemannn) [inline ios/Classes/FacebookLikeButton.m:94]: The attributes "center", "left" and "right" not seem to be documented.

File: ios/Classes/FacebookLikeButton.m [hunk 7]
```diff
@@ -50,63 +55,56 @@ -(void)setForegroundColor_:(id)colorVal
...
-    FBLikeControl *btn = [self like];
-    NSString* oType = [TiUtils stringValue:typeStr];
-    if ([oType isEqualToString:@"page"]) {
-        btn.objectType = FBLikeControlObjectTypePage;
-    } else if ([oType isEqualToString:@"openGraphObject"]) {
-        btn.objectType = FBLikeControlObjectTypeOpenGraphObject;
-    } else {
-        btn.objectType = FBLikeControlObjectTypeUnknown;
-    }
+    DEPRECATED_REMOVED(@"facebook.likeButton.objectType", @"5.0.0", @"5.0.0");
```
Comment (hansemannn) [inline ios/Classes/FacebookLikeButton.m]: Capitalize "facebook" to "Facebook".

File: ios/Classes/FacebookModule.m [hunk 8]
```diff
@@ -286,32 +181,66 @@ -(id)accessToken
...
     }, YES);
-    
     return token;
 }
-
+//deprecated and removed
 -(id)AUDIENCE_NONE
 {
-    return [NSNumber numberWithInt:FBSessionDefaultAudienceNone];
+    DEPRECATED_REMOVED(@"facebook.AUDIENCE_NONE",@"5.0.0",@"5.0.0")
```
Comment (hansemannn) [inline ios/Classes/FacebookModule.m]: Capitalize "facebook" to "Facebook".

File: ios/Classes/FacebookModule.m [hunk 9]
```diff
@@ -286,32 +181,66 @@ -(id)accessToken
 {
     __block NSString * token;
     TiThreadPerformOnMainThread(^{
-        token = FBSession.activeSession.accessTokenData.accessToken;
+		token = [[FBSDKAccessToken currentAccessToken] tokenString];
```
Comment (hansemannn) [inline ios/Classes/FacebookModule.m:184]: Indentation level seems shifted, please check in all cases.

File: ios/Classes/FacebookModule.m [hunk 10]
```diff
@@ -394,17 +322,19 @@ -(void)authorize:(id)args
...
-    }, NO);
+        [loginManager logInWithReadPermissions: permissions_ fromViewController:nil handler:^(FBSDKLoginManagerLoginResult *result, NSError *error) {
+            if (error) {
+                DebugLog(@"[ERROR] Process error.");
+            } else if (result.isCancelled) {
+                DebugLog(@"[ERROR] User Cancelled");
+            } else {
+                DebugLog(@"[INFO] Logged in");
+            }
+        }];
```
Comment (hansemannn) [inline ios/Classes/FacebookModule.m:338]: Should this logs remain there? If so, unify the sentences with a closing "." and check the lowercase spelling of "Cancelled".

File: ios/Classes/FacebookModule.m [hunk 11]
```diff
@@ -413,21 +343,18 @@ -(void)authorize:(id)args
 -(void)initialize:(id)args
 {
     ENSURE_SINGLE_ARG_OR_NIL(args, NSNumber);
-    double timeoutInMs = [TiUtils intValue:args def:0];
-    if (timeoutInMs == 0) {
-        meRequestTimeout = 180.0;
-    } else {
-        meRequestTimeout = (double)timeoutInMs / 1000.0;
+    if (args != nil) {
+        DEPRECATED_REMOVED(@"timeout param", @"5.0.0", @"5.0.0");
```
Comment (hansemannn) [inline ios/Classes/FacebookModule.m]: Considering, if "timeout param" is the right name to conform with the log output.

File: apidoc/LikeButton.yml (hunk 1 above)
Comment (cheekiatng) [inline apidoc/LikeButton.yml:118]: this hasn't been removed. just deprecated. will be removed later.

File: README.md (hunk 2 above)
Comment (cheekiatng) [inline README.md:198]: yup. facebook API will determine if the dialog is presented in the Facebook native app or the web browser.

File: ios/Classes/FacebookLikeButton.m (hunk 5 above)
Comment (cheekiatng) [inline ios/Classes/FacebookLikeButton.m:74]: it's here: https://github.com/appcelerator-modules/ti.facebook/blob/master/apidoc/LikeButton.yml#L77

File: ios/Classes/FacebookLikeButton.m (hunk 6 above)
Comment (cheekiatng) [inline ios/Classes/FacebookLikeButton.m:94]: here: https://github.com/appcelerator-modules/ti.facebook/blob/master/apidoc/LikeButton.yml#L95

File: ios/Classes/FacebookModule.m (hunk 10 above)
Comment (cheekiatng) [inline ios/Classes/FacebookModule.m:338]: Noted. thanks for this, i also missed the case of a cancelled event. PR coming.

File: apidoc/LikeButton.yml (hunk 1 above)
Comment (cheekiatng) [inline apidoc/LikeButton.yml:118]: Never mind. I'll remove it now since it's a major version bump.

File: ios/Classes/FacebookModule.m (hunk 11 above)
Comment (cheekiatng) [inline ios/Classes/FacebookModule.m]: that timeout parameter is for internal use and not public API

**Comment by cheekiatng** [comment]
Will capitalise facebook to Facebook

**Comment by hansemannn** [comment]
@cheekiatng When does the user need to call fb.initialize() after require'ing the module? In the README.md, it is used sometimes and sometimes not.

Also getting this error when trying to read the user's group `fb.requestWithGraphPath('me/groups', {}, 'GET',  function(r){})`

```
FBSDKLog: starting with Graph API v2.4, GET requests for /me/groups should contain an explicit "fields" parameter
```

And: When I try to use the like button from the simulator (with web auth), I get this one: `Presenting view controllers on detached view controllers is discouraged <TiViewController: 0x7fb141775860>.`

**Comment by cheekiatng** [comment]
fb.initialize() is required to be called at least once in an app, before anything in the fb module can be used. 
The second error implies that there was an issue with the Graph API Request. That, we have to check out Facebook what's the new requirements for this request, i'm guessing it's exactly what it's saying, it needs a "fields" parameter.
3rd part has nothing to do with the module. it's the way Kitchensink was coded.

**Comment by hansemannn** [comment]
Thanks for clarifying that!

**Comment by hansemannn** [comment]
FT and CR passed, PR **approved**. Thank you @cheekiatng !

**Comment by bogdansmc** [comment]
As it is indicated here: https://github.com/appcelerator/titanium_mobile/pull/7292
This module doesn't work and crash app.

**Pull request merged**